### ISSUE TITLE
Queries hanging on tables being actively streamed to

### ISSUE THREAD

**Issue by JonathonO** [snowflakedb/snowflake-kafka-connector#548]
Confluent Kafka 7.3.1-ce
Snowflake Kafka Connector 1.8.3

Following Snowflake PrPr (also enabled on the destination Snowflake account) options configured in the Sink:
```
    ...
    "snowflake.enable.schematization": "true",
    "snowflake.ingestion.method": "SNOWPIPE_STREAMING",
    "snowflake.schematization.auto": "false",
    ...
```

Destination tables are created and receiving data from the Sink. `SHOW TABLES` shows rows and bytes increasing and most are queryable.

For tables actively receiving data and being queried simultaneously, the queries seem to hang.

When we remove the affected table from the Sink configuration though, queries on that table return results. For example, a table with approx. 7 million rows, `SELECT COUNT(column) FROM table` returns in 0-1 seconds. We repeated this several times and got the same result.

Although it's using `SNOWPIPE_STREAMING`, we did check `SHOW LOCKS` and `SHOW TRANSACTIONS` for the Snowflake database user used by the Sink while the affected table was receiving data and both showed 0 results.

Checked Snowflake Status Page also, and Snowpipe Status for the Snowflake account's region (`AWS_US_EAST_1`) showing as operational but, not sure if the PrPr features would be covered by your status page or not?

Failed Query ID example: `01aa86db-0603-1053-119a-030100a40adb`. We terminated it after it hung for over 40 seconds.

Please let us know what other information we can get to diagnose this.

**Comment by JonathonO** [comment]
So it seems to be intermittent. Further testing with simple aggregation queries showed some queries hanging for minutes (cancelled by user), others returning within seconds.

Here's some examples from today with total elapsed time for a simple `SELECT COUNT(id) FROM pubs` query.
```
01aa8ae2-0603-1053-119a-0301012bc79b | 1351   | SUCCESS
01aa8adf-0603-109f-119a-0301012b1a2f | 194758 | FAILED_WITH_ERROR | 604 Query execution cancelled
01aa8adf-0603-109f-119a-0301012b19cf | 80     | SUCCESS
01aa8add-0603-1053-119a-0301012bc2ff | 19240  | SUCCESS
```

**Comment by sfc-gh-tjones** [comment]
Hey @JonathonO do you have an account URL handy so I can take a look at the queries? Thanks!

**Comment by JonathonO** [comment]
Hi @sfc-gh-tjones yes, I do. What's the best way to share it with you?

**Comment by sfc-gh-tjones** [comment]
Actually I found it :) . For `01aa8adf-0603-109f-119a-0301012b1a2f` what looks like happened is that when the query ran it hit a compute node that did not have metadata cached for that table so it had to do a load into a cold cache. Furthermore, it looks like there were many small data files it was fetching metadata for so this took some additional time. I would recommend buffering data if you can before handing it off to the SDK so that it generates fewer small files which would result in less files needing to be considered by our compiler. We have some optimizations on our side planned to mitigate this but in the short term this is what I'd recommend, although note eventually these queries will complete once the cache is populated.

**Comment by JonathonO** [comment]
@sfc-gh-tjones I'm sure we can do that. Is there a preferred or ideal buffer size for these files? Doesn't have to be precise, obviously.

**Comment by JonathonO** [comment]
@sfc-gh-tjones just to keep us all in the loop, I've been told directly that the issue relates to internal file metadata and compaction, and that dropping and recreating the table was recommended - which we've done. Seems to align with what you're saying also :)

**Comment by sfc-gh-tjones** [comment]
I would say if you can get between 1-16 megabytes that would be ideal. The closer toward 16 the better.

Yes, this aligns. Compaction consolidates metadata so its faster to do these types of queries until after that takes place.

**Comment by JonathonO** [comment]
@sfc-gh-tjones we've released Snowflake Sink v1.9.1 today to resolve these other issues #553 #558. Do you think we should still configure the Consumer to buffer/batch records? If so, is dropping and recreating the table still required?

**Comment by sfc-gh-tjones** [comment]
No need to drop/recreate the table: I would buffer/batch records to guarantee the lower count(*) latency, otherwise you may run into queries that hang although that should still be rare. Note that these queries should ALWAYS complete, there's not a fundamental issue preventing us from returning that result.
